File: 7/main.py
from collections import defaultdict
from typing import List


class FSItem():
    def __init__(self, path: str, size=0, kind='dir') -> None:
        self.path = path
        self.size = size
        self.files: list[FSItem] = []
        self.kind = kind

    def add_dir(self, path, name):
        print("add dir")
        print(f"{path}")
        print(f"{name}")
        print(f"{path.split('/')}")
        current, *rest = path.split("/")
        rest_str = "/".join(rest)
        print(f"current: {current}")
        print(f"rest: {rest}")
        print(f"rest_str: {rest_str}")
        if not rest_str:
            self.files.append(FSItem(name))
        else:
            list(filter(lambda x: x.path == rest[0], self.files))[0].add_dir(rest_str, name)

    def add_file(self, path, name, size):
        print("add file")
        current, *rest = path.split("/")
        rest_str = "/".join(rest)
        print("rest_str")
        print(rest_str)
        print("name")
        print(name)
        print("current")
        print(current)
        if not rest_str:
            self.files.append(FSItem(name, int(size), kind='file'))
        else:
            list(filter(lambda x: x.path == rest[0], self.files))[0].add_file(rest_str, name, size)

    def update_size(self):
        result = 0
        for file in self.files:
            if file.size:
                result += file.size
            else:
                file.update_size()
                result += file.size
        self.size = result
        return result

    def __str__(self) -> str:
        return f"FSItem({self.path}, {self.size}, {[str(file) for file in self.files]})"

    def __radd__(self, other):
        return other + self.size

    def __lt__(self, other):
        return self.size < other.size

    def children(self):
        result = [self]
        for file in self.files:
            result.extend(file.children()) 
        return result

    def to_string(self, n=0) -> str:
        new_line = "\n"
        files_str = "".join([file.to_string(n+1) for file in self.files])
        return f"{'  '*n}- {self.path} ({self.kind}{', ' + str(self.size) if self.size else ''}){new_line}{files_str}"




CD = 'cd'
LS = 'ls'

f = open("7/input.txt")
comand_prefix = "$ "
current_dir = '/'
last_command = None

fs = FSItem('/')
for l in f.readlines():
    print(fs.to_string())
    l = l.strip()
    print(f"l: {l}")
    print(f"current_dir: {current_dir}")
    if l.startswith(comand_prefix):
        comand = l.removeprefix(comand_prefix)
        if comand.startswith(CD):
            last_command = CD
            program, args = comand.split(" ")
            if args == "..":
                print('before')
                print(current_dir)
                print(current_dir.split("/"))
                current_dir = "/".join(current_dir.split('/')[:-2]) + "/"
                print('after')
                print(current_dir)
            elif args == "/":
                current_dir = args
            else:
                print('will set current_dir')
                print(args)
                current_dir += f"{args}/"
                print(current_dir)
        elif comand.startswith(LS):
            last_command = LS
        else:
            raise ValueError(f'error parsing program: {l}')
    else:
        if last_command == LS:
            if l.startswith("dir"):
                _, name = l.split("dir ")
                fs.add_dir(current_dir, name)
                print('added files[] on current_dir{current_dir}')
            else:
                size, name = l.split(" ")
                fs.add_file(current_dir, name, size)
        elif last_command == CD:
            pass
        else:
            raise ValueError(f'error parsing input: {l}')
print(fs)
print(fs.to_string())
print(fs.update_size())
print(fs.to_string())
print([str(f) for f in fs.children()])
part1: List[FSItem] = list(filter(lambda x: x.size < 100000 and x.kind == 'dir', fs.children()))
print([str(f) for f in part1])
print(sum(part1))
max = fs.size
print(f"max:{max}")
unused = 70000000 - max
print(f"unused:{unused}")
needed_to_free_up = 30000000 - unused
print(f"needed_to_free_up:{needed_to_free_up}")
part2: List[FSItem] = list(filter(lambda x: x.size > needed_to_free_up and x.kind == 'dir', fs.children()))
# print([str(f) for f in part2])
part2.sort()
# print([str(f) for f in part2])
print("min(part2)")
print(min(part2).size)

f.close()

File: 7/src/main.rs
use std::env;
use std::fs;
use std::result;

#[derive(Debug, Clone, PartialEq)]
enum Kind {
    Dir,
    File,
}
#[derive(Debug, Clone)]
struct FSItem {
    path: String,
    size: i32,
    kind: Kind,
    files: Vec<Box<FSItem>>,
}

impl FSItem {
    fn add_dir(&mut self, path: &str, name: &str) {
        let (_, rest) = path.split_once("/").unwrap();
        println!("{:?}", rest);
        if rest.len() > 0 {
            let (next, _) = rest.split_once("/").unwrap();
            let file = self
                .files
                .iter()
                .filter(|f| f.path == next)
                .map(|fs| fs)
                .last()
                .unwrap();
            println!("{:?}", file);
            let index = self.files.iter().position(|f| f.path == file.path).unwrap();
            let mut cloned_file = file.clone();
            cloned_file.add_dir(rest, name);
            self.files.remove(index);
            self.files.extend([cloned_file]);
        } else {
            self.files.extend([Box::new(FSItem {
                path: name.to_string(),
                kind: Kind::Dir,
                files: vec![],
                size: 0,
            })])
        }
    }
    fn add_file(&mut self, path: &str, name: &str, size: i32) {
        let (_, rest) = path.split_once("/").unwrap();
        if rest.len() > 0 {
            let (next, _) = rest.split_once("/").unwrap();
            let file = self
                .files
                .iter()
                .filter(|f| f.path == next)
                .map(|fs| fs)
                .last()
                .unwrap();
            println!("{:?}", file);
            let index = self.files.iter().position(|f| f.path == file.path).unwrap();
            let mut cloned_file = file.clone();
            cloned_file.add_file(rest, name, size);
            self.files.remove(index);
            self.files.extend([cloned_file]);
        } else {
            self.files.extend([Box::new(FSItem {
                path: name.to_string(),
                kind: Kind::File,
                files: vec![],
                size: size,
            })])
        }
    }
    fn update_size(&mut self) -> i32 {
        let mut result = 0;
        for file in &mut self.files {
            if file.size != 0 {
                result += file.size;
            } else {
                file.update_size();
                result += file.size;
            }
        }
        self.size = result;
        return result;
    }
    fn children(self) -> Vec<FSItem> {
        let clone = self.clone();
        let mut result = vec![self];
        let files = clone.files.clone();
        for file in files {
            result.extend(file.children())
        }
        return result;
    }
}

fn main() {
    let mut current_dir = env::current_dir().unwrap().to_str().unwrap().to_owned();
    let today = "/7";
    if !current_dir.ends_with(today) {
        current_dir += today
    }
    let contents = fs::read_to_string("7/input.txt").expect("couldn't read file");
    let lines = contents.lines();
    let mut fs = FSItem {
        path: "/".to_string(),
        kind: Kind::Dir,
        files: vec![],
        size: 0,
    };

    let comand_prefix = "$ ";
    let mut current_dir = "/".to_owned();
    for mut l in lines.into_iter() {
        println!("{:?}", l);
        l = l.trim();
        if l.starts_with(comand_prefix) {
            let comand = l.strip_prefix(comand_prefix).unwrap();
            if comand.starts_with("cd") {
                let (_, args) = comand.split_once(" ").unwrap();
                if args == ".." {
                    let current_dir_size = current_dir.matches("/").count();
                    current_dir = current_dir
                        .split_inclusive("/")
                        .into_iter()
                        .enumerate()
                        .filter(|&(i, _)| i != current_dir_size - 1)
                        .map(|(_, e)| e)
                        .collect();
                } else if args == "/" {
                    current_dir = args.to_string();
                } else {
                    current_dir = current_dir + args + "/";
                }
            } else if comand.starts_with("ls") {
            }
        } else {
            if l.starts_with("dir ") {
                let (_, name) = l.split_once("dir ").unwrap();
                fs.add_dir(&current_dir, name);
            } else {
                let (size, name) = l.split_once(" ").unwrap();
                fs.add_file(&current_dir, name, size.parse::<i32>().unwrap());
            }
        }
    }

    println!("{:?}", fs);
    fs.update_size();
    println!("{:?}", fs);
    let max = fs.size;
    let part2fs = fs.clone();
    let part1: i32 = fs
        .children()
        .into_iter()
        .filter(|f| f.size < 100000 && f.kind == Kind::Dir)
        .map(|f| f.size)
        .sum();
    println!("Part1: {:?}", part1);
    // part2
    let unused = 70000000 - max;
    let needed_to_free_up = 30000000 - unused;
    let part2 = part2fs
        .children()
        .into_iter()
        .filter(|f| f.size > needed_to_free_up && f.kind == Kind::Dir)
        .map(|f| f.size)
        .min()
        .unwrap();
    println!("Part2: {:?}", part2);
}
